packages/flutter_driver/lib/src/driver/profiling_summarizer.dart

```diff
@@ -0,0 +1,160 @@
+/// for supported profiles, i.e, CPU, Memory and GPU. Not all of these exist for
+/// all the platforms.
+class ProfilingSummarizer {
+  ProfilingSummarizer._(this.eventByType);
+
+  /// Creates a ProfilingSummarizer given the timeline events.
+  static ProfilingSummarizer fromEvents(List<TimelineEvent> profilingEvents) {
+    final Map<ProfileType, List<TimelineEvent>> eventByType =
+        <ProfileType, List<TimelineEvent>>{};
+    for (final TimelineEvent event in profilingEvents) {
+      assert(event.category == kProfilingCategory);
+      final ProfileType type = _getProfileType(event.name);
+      if (eventByType.containsKey(type)) {
+        eventByType[type].add(event);
+      } else {
+        eventByType[type] = <TimelineEvent>[event];
+      }
+    }
+    return ProfilingSummarizer._(eventByType);
+  }
+
+  /// Key is the type of profiling event, for e.g. CPU, GPU, Memory.
+  final Map<ProfileType, List<TimelineEvent>> eventByType;
+
+  /// Returns the average, 90th and 99th percentile summary of CPU, GPU and Memory
+  /// usage from the recorded events. Note: If a given profile type isn't available
+  /// for any reason, the map will not contain the said profile type.
+  Map<String, dynamic> summarize() {
+    final Map<String, dynamic> summary = <String, dynamic>{};
+    summary.addAll(_summarize(ProfileType.CPU, 'cpu_usage'));
+    summary.addAll(_summarize(ProfileType.GPU, 'gpu_usage'));
+    summary.addAll(_summarize(ProfileType.Memory, 'memory_usage'));
+    return summary;
+  }
+
+  Map<String, dynamic> _summarize(ProfileType profileType, String name) {
+    final Map<String, dynamic> summary = <String, dynamic>{};
+    if (!hasProfilingInfo(profileType)) {
+      return summary;
+    }
+    summary['average_$name'] = computeAverage(profileType);
+    summary['90th_percentile_$name'] = computePercentile(profileType, 90);
+    summary['99th_percentile_$name'] = computePercentile(profileType, 99);
+    return summary;
+  }
+
+  /// Returns true if there are events in the timeline corresponding to [profileType].
+  bool hasProfilingInfo(ProfileType profileType) {
+    if (eventByType.containsKey(profileType)) {
+      return eventByType[profileType].isNotEmpty;
+    } else {
+      return false;
+    }
+  }
+
+  /// Computes the average of the `profileType` over the recorded events.
+  double computeAverage(ProfileType profileType) {
+    final List<TimelineEvent> events = eventByType[profileType];
+    if (events.isEmpty) {
+      return 0;
+    }
+
+    final double total = events
+        .map((TimelineEvent e) => _getProfileValue(profileType, e))
+        .reduce((double a, double b) => a + b);
+    return total / events.length;
+  }
+
+  /// The [percentile]-th percentile `profileType` over the recorded events.
+  double computePercentile(ProfileType profileType, double percentile) {
+    final List<TimelineEvent> events = eventByType[profileType];
+    if (events.isEmpty) {
```

Comment:
Returning 0 if the list is empty may have the following problems:
- It's unclear whether the 0 is caused by an empty list or that all values are 0 in the list.
- It's going to add many 0-valued metrics to our benchmarks that currently don't have CPU or memory measurements (i.e., all Android tests). It will add a lot of dummy graphs to our dashboard, and may need a lot of baselining work.

Maybe we should assert the list is not empty when we're expecting CPU/memory measurements, and drop the value (not calling computePercentile) when such measurement doesn't exist.

Reply:
The value is already dropped when it doesn't exist. I do a check here: https://github.com/flutter/flutter/pull/58820/files#diff-9277fbf8cff9518e455fc51bac717a7aR81 to ensure that a profile exists. I will change this to an assert though, given that this should never happen.